plugins/inputs/windows_event_log/wineventlog/wineventlog_test.go

```diff
@@ -0,0 +1,163 @@
+)
+
+var (
+	// common inputs for creating an EventLog.
+	NAME = "Application"
+	// 2 is ERROR
+	LEVELS          = []string{"2"}
+	GROUP_NAME      = "fake"
+	STREAM_NAME     = "fake"
+	RENDER_FMT      = FormatPlainText
+	DEST            = "fake"
+	STATE_FILE_PATH = "fake"
+	BATCH_SIZE      = 99
+	RETENTION       = 42
+)
+
+// TestNewEventLog verifies constructor's default values.
+func TestNewEventLog(t *testing.T) {
+	elog := NewEventLog(NAME, LEVELS, GROUP_NAME, STREAM_NAME, RENDER_FMT, DEST,
+		STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.Equal(t, NAME, elog.name)
+	assert.Equal(t, uint64(0), elog.eventOffset)
+	assert.Zero(t, elog.eventHandle)
+}
+
+// TestOpen verifies Open() succeeds with valid inputs.
+// And fails with invalid inputs.
+func TestOpen(t *testing.T) {
+	// happy
+	elog := NewEventLog(NAME, LEVELS, GROUP_NAME, STREAM_NAME, RENDER_FMT, DEST,
+		STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.NoError(t, elog.Open())
+	assert.NotNil(t, elog.eventHandle)
+	assert.NoError(t, elog.Close())
+	// bad name
+	elog = NewEventLog("FakeBadElogName", LEVELS, GROUP_NAME, STREAM_NAME,
+		RENDER_FMT, DEST, STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.Error(t, elog.Open())
+	assert.Zero(t, elog.eventHandle)
+	// bad LEVELS does not cause Open() to fail.
+	// bad wlog.eventOffset does not cause Open() to fail.
+}
+
+// TestReadGoodSource will verify we can read events written by a registered
+// event log source.
+func TestReadGoodSource(t *testing.T) {
+	elog := NewEventLog(NAME, LEVELS, GROUP_NAME, STREAM_NAME, RENDER_FMT, DEST,
+		STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.Equal(t, nil, elog.Open())
+	seekToEnd(t, elog)
+	writeEvents(t, 10, true, "CWA_UnitTest111", 777)
+	// Without a sleep we do not read any events!!!
+	time.Sleep(1 * time.Second)
+	checkEvents(t, elog, 10, "[Application] [ERROR] [777] [CWA_UnitTest111] ")
+	assert.Equal(t, nil, elog.Close())
+}
+
+// TestReadBadSource will verify that we cannot read events written by an
+// unregistered event log source.
+func TestReadBadSource(t *testing.T) {
+	elog := NewEventLog(NAME, LEVELS, GROUP_NAME, STREAM_NAME, RENDER_FMT, DEST,
+		STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.Equal(t, nil, elog.Open())
+	seekToEnd(t, elog)
+	writeEvents(t, 10, false, "CWA_UnitTest222", 888)
+	time.Sleep(1 * time.Second)
+	checkEvents(t, elog, 0, "[Application] [ERROR] [888] [CWA_UnitTest222] ")
+	assert.Equal(t, nil, elog.Close())
+}
+
+// TestReadWithBothSources will verify we can read events written by a
+// registered event log source, even if the batch contains events from an
+// unregistered source too.
+func TestReadWithBothSources(t *testing.T) {
+	elog := NewEventLog(NAME, LEVELS, GROUP_NAME, STREAM_NAME, RENDER_FMT, DEST,
+		STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.Equal(t, nil, elog.Open())
+	seekToEnd(t, elog)
+	writeEvents(t, 10, true, "CWA_UnitTest111", 777)
+	writeEvents(t, 10, false, "CWA_UnitTest222", 888)
+	time.Sleep(1 * time.Second)
```

Comment:
What do we need the sleep for? Just going to leave this one comment, but I'm applying it to all of the sleep statements in the tests. I did not have fun with the tailersrc tests and their flaky timing. Don't want a repeat of that here if we can avoid it.

Reply:
Because of a bad behavior in this package:
`"golang.org/x/sys/windows/svc/eventlog"`
https://cs.opensource.google/go/x/sys/+/1d35b9e2:windows/svc/eventlog/log.go;l=52

I added this comment in 1 place, but not all.
```
	// Without a sleep we do not read any events!!!
	// To work around the bad behavior of writeEvents() we must sleep.
	time.Sleep(1 * time.Second)
```
Let me know if you prefer I add that comment before ALL sleep() calls.

I definitely tested without the sleep and it failed.
I considered polling until we get the new events we expect or reach a time limit.
A single sleep call is simpler.

Reply:
I would prefer adding the comment on all of them, but that's just me. That way, as things move around/change, we still have reference to this issue in every test that has the sleep to account for it.

Reply:
Fixed.